### PR TITLE
I1181 diag plots

### PR DESCRIPTION
**Pull request by rcw5890**
I added implementations of the noise model diagnostics @ben18785 and I have been discussing, for #1181

### CODE REVIEW

File: docs/source/noise_model_diagnostics.rst [hunk 1]
```diff
@@ -10,6 +10,9 @@ Plotting functions:
 
 - :func:`plot_residuals_autocorrelation`
 - :func:`plot_residuals_vs_output`
+- :func:`plot_residuals_binned_autocorrelation`
```
Comment (ben18785) [inline docs/source/noise_model_diagnostics.rst:13]: Can we make these alphabetic please?

File: docs/source/noise_model_diagnostics.rst [hunk 2]
```diff
@@ -24,6 +27,12 @@ Plotting functions
 
 .. autofunction:: plot_residuals_vs_output
 
+.. autofunction:: plot_residuals_binned_autocorrelation
```
Comment (ben18785) [inline docs/source/noise_model_diagnostics.rst:30]: Alphabetic too please

File: examples/README.md [hunk 3]
```diff
@@ -72,6 +72,8 @@ relevant code.
 - [Autocorrelation](./plotting/mcmc-autocorrelation.ipynb)
 - [Effective sample size](./sampling/effective-sample-size.ipynb)
 - [Evaluating noise models using autocorrelation plots of the residuals](./plotting/residuals-autocorrelation.ipynb)
+- [Noise model diagnostic plots (correlation)](./plotting/residuals-autocorrelation-diagnostics.ipynb)
```
Comment (ben18785) [inline examples/README.md:75]: Can these be alphabetically ordered within this list?

File: pints/residuals_diagnostics.py [hunk 4]
```diff
@@ -11,6 +11,311 @@
 import scipy.special
 
 
+def plot_residuals_binned_autocorrelation(parameters,
+                                          problem,
+                                          thinning=None,
+                                          n_bins=25):
+    r"""
+    Plot the autocorrelation of the residuals within bins.
```
Comment (ben18785) [inline pints/residuals_diagnostics.py:19]: Can we change this to "Plot the autocorrelation of the residuals within bins (i.e. discrete time windows across the series)".

File: pints/residuals_diagnostics.py [hunk 5]
```diff
@@ -11,6 +11,311 @@
...
+
+    .. math::
+        e_i = y_i - f(t_i; \theta)
+
+    This method divides the vector of residuals into some number of equally
+    sized bins. The lag 1 autocorrelation is calculated for the residuals
+    within each bin. The plot shows the lag 1 autocorrelation in each bin over
+    time.
+
+    This diagnostic is particularly useful for diagnosing time series whose
```
Comment (ben18785) [inline pints/residuals_diagnostics.py:31]: I would phrase this as, "This diagnostic is useful for diagnosing time series with noise whose autocorrelation varies over time."

File: pints/residuals_diagnostics.py [hunk 6]
```diff
@@ -11,6 +11,311 @@
...
+    noise exhibits autocorrelation at certain time periods.
+
+    When passing an array of parameters (from an MCMC sampler), this method
+    will plot the autocorrelations of the posterior median residual values.
```
Comment (ben18785) [inline pints/residuals_diagnostics.py:35]: "this method plots"

File: pints/residuals_diagnostics.py [hunk 7]
```diff
@@ -11,6 +11,311 @@
...
+
+    Typically, this diagnostic can be called after obtaining the residuals of
```
Comment (ben18785) [inline pints/residuals_diagnostics.py:37]: "Typically, this diagnostic is called after obtaining..."

File: pints/residuals_diagnostics.py [hunk 8]
```diff
@@ -11,6 +11,311 @@
...
+    .. math::
+        e_i = y_i - f(t_i; \theta)
+
+    this function generates and plots the distance matrix :math:`D` whose
+    entries are defined by
+
+    .. math::
+        D_{i, j} = |e_i - e_j|
+
+    The plot of this matrix may be helpful for identifying a time series with
```
Comment (ben18785) [inline pints/residuals_diagnostics.py:248]: Isn't it more that banded noise structure suggests that autocorrelation is temporally varying?

File: pints/residuals_diagnostics.py [hunk 9]
```diff
@@ -11,6 +11,311 @@
...
+    correlated noise. When the noise terms are correlated, the distance matrix
+    :math:`D` is likely to have a banded appearance.
+
+    For problems with multiple outputs, one distance matrix is generated for
+    each output.
+
+    When passing an array of parameters (from an MCMC sampler), this method
+    will plot the distance matrix of the posterior median residual values.
+
+    Typically, this diagnostic can be called after obtaining the residuals of
```
Comment (ben18785) [inline pints/residuals_diagnostics.py:258]: "Typically, this diagnostic is called after obtaining the residuals of..."

**Comment by rcw5890** [comment]
> Thanks @rcw5890 -- looks good! Apart from my minor comments on the docstrings, all looks good. My one slightly bigger comment would be:
> 
> * In the variance diagnostics notebook, it's not obvious to me from the plot of output vs fitted that the standard deviation is changing over time? Perhaps my eyes aren't tuned to the example but the distance of points from the line doesn't seem to be varying dramatically over the series...
> * In the same notebook, can we turn off the print to screen thing for the optimiser?

I'll take a look at this notebook again. I think I can get a better visual impact by adjusting the various parameters.

File: pints/residuals_diagnostics.py (hunk 8 above)
Comment (rcw5890) [reply pints/residuals_diagnostics.py:250]: I think I can improve the wording here. But wont the matrix still appear banded even for a level of autocorrelation which is constant over time? I'm thinking of one example I did in the [notebook](https://github.com/pints-team/pints/blob/b6e8c60d6e1d80a71cda55acef2b734f28768e4f/examples/plotting/residuals-autocorrelation-diagnostics.ipynb) (cell 6), where the distance matrix still appears distinctively banded for AR(1) noise with constant rho=0.85.

**Comment by rcw5890** [comment]
For the variance diagnostics example, I tried some different parameters for logistic growth and the multiplicative noise scaling. The latest version should show the temporally varying noise magnitude more clearly.

**Review by ben18785 — APPROVED** [review]
Looks great, thanks @rcw5890

**Pull request merged**